db/virtual_table.cc

```diff
@@ -61,4 +62,43 @@ bool virtual_table::contains_key(const dht::partition_range& pr, const dht::deco
     return pr.contains(dk, dht::ring_position_comparator(*_s));
 }
 
+mutation_source memtable_filling_virtual_table::as_mutation_source() {
+    return mutation_source([this] (schema_ptr s,
+        reader_permit permit,
+        const dht::partition_range& range,
+        const query::partition_slice& slice,
+        const io_priority_class& pc,
+        tracing::trace_state_ptr trace_state,
+        streamed_mutation::forwarding fwd,
+        mutation_reader::forwarding fwd_mr) {
+
+        auto units = make_lw_shared<>(permit.consume_memory(0));
+
+        auto populate = [this, mt = make_lw_shared<memtable>(schema()), s, units, range, slice, pc, trace_state, fwd, fwd_mr] (db::timeout_clock::time_point timeout) mutable {
+            auto mutation_sink = [units, mt, memory_used = uint64_t(0)] (mutation m) mutable {
```

Comment:
This lambda can be copied around by the user code. std::function<>, in which it is wrapped, will copy it when it is copied. This can create multiple instances of `memory_used`. One way to fix that is to share it under `units`, turning it into a struct.

Reply:
@tgrabiec done